File: sim-diasca/src/core/services/dataflow/bindings/python/api/sim_diasca/dataflow_types.py
# Copyright (C) 2016-2017 EDF R&D
#
# This file is part of Sim-Diasca.
#
# Sim-Diasca is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of
# the License, or (at your option) any later version.
#
# Sim-Diasca is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with Sim-Diasca.
# If not, see <http://www.gnu.org/licenses/>.

# Author: Robin Huart (robin-externe.huart@edf.fr)


# Standard utilities:
from enum import Enum, unique
from typing import List, Union, Tuple


# Currently, workbench-related developments rely on a Python virtual
# environment, whereas the engine itself directly imports modules.

try:
    # If run directly from the engine:
    from common.trace_emitter import TraceEmitter

except ImportError:
    # If run from a Python virtual environment:
    from .common.trace_emitter import TraceEmitter


##
##
# --- ENUMERATIONS ---
##
##


# Base class defining enums with automatic numbering and hidden values:
#
# Note: inspired by the official Python documentation:
#           https://docs.python.org/3/library/enum.html
#
class AutoHidingEnum(Enum):
    """
    Base class defining enums with automatic numbering and hidden values.
    """

    def __new__(cls):
        value = len(cls.__members__) + 1
        obj = object.__new__(cls)
        obj._value_ = value
        return obj

    def __repr__(self):
        return '<{}.{}>'.format(self.__class__.__name__, self.name)

    def encode(self):
        lower_case_name = self.name.lower()
        return lower_case_name.encode()


# The activation policy governing the condition(s) to match before triggering
# the evaluation of the 'activate' method of a processing unit:
#
@unique
class ActivationPolicy(AutoHidingEnum):
    """Enumeration matching the activation policies available in the core of the
    engine.
    """
    ACTIVATE_ON_NEW_SET = ()
    ACTIVATE_WHEN_ALL_SET = ()
    CUSTOM_ACTIVATION = ()


# The available constraints, against which any port value can be checked:
#
# Note: these must match the list of constraints in Sim-Diasca, which can be
# found in class_DataflowBlock_functions.hrl:validate_constraints/2 as Erlang
# atoms.
#
@unique
class PortConstraints(AutoHidingEnum):
    """
    Enumeration matching the constraints available in Sim-Diasca.
    """
    NON_NULL = ()
    STRICTLY_NEGATIVE = ()
    NEGATIVE = ()
    STRICTLY_POSITIVE = ()
    POSITIVE = ()
    IN = ()                      # to be used as ( IN, list )
    BETWEEN = ()                 # to be used as ( BETWEEN, x, y )
    LOWER_THAN = ()              # to be used as ( LOWER_THAN, x )
    GREATER_THAN = ()            # to be used as ( GREATER_THAN, x )


##
##
# --- CLASSES ---
##
##


# Definition of the type used for specifying the input port(s) of a processing
# unit:
#
class InputPortSpecification:
    """
    User-defined specifications describing an input port (iterated or not).
    """

    # Predefined sequence of attributes:
    __slots__ = ('name', 'comment', 'is_iteration', 'value_semantics',
                 'value_unit', 'value_type_description', 'value_constraints')

    # Initial definition of an input port specification:
    def __init__(self, name, comment, value_semantics, value_unit,
                 value_type_description, value_constraints=[],
                 is_iteration=False):
        """
        Initial specification of all the characteristics of an input port.
        """

        self.name = name
        self.comment = comment
        self.is_iteration = is_iteration
        self.value_semantics = value_semantics
        self.value_unit = value_unit
        self.value_type_description = value_type_description
        self.value_constraints = value_constraints

    # Encoding funtion:
    def encode(self) -> list:
        """
        Transforms an InputPortSpecification object into a list of tuple pairs,
        more easily handled by ErlPort.
        """

        encoded_semantics = [semstring.encode() for semstring
                             in self.value_semantics]

        encoded_constraints = []
        # each of the constraints might be encoded in different ways
        for constraint in self.value_constraints:

            encoded_constraint = None

            if (isinstance(constraint, tuple) and
                    len(constraint) == 2 and
                    isinstance(constraint[1], list)):
                # it might be a tuple made of a string and list, such as ("IN",
                # ["a","b"])
                encoded_constraint = (constraint[0].encode(), [
                                      v.encode() for v in constraint[1]])
            else:
                # by default we encode anything we find (hope it works)
                encoded_constraint = constraint.encode()

            encoded_constraints.append(encoded_constraint)

        return [(b'name', self.name.encode()),
                (b'comment', self.comment.encode()),
                (b'is_iteration', self.is_iteration),
                (b'value_semantics', encoded_semantics),
                (b'value_unit', self.value_unit.encode()),
                (b'value_type_description',
                 self.value_type_description.encode()),
                (b'value_constraints', encoded_constraints)]


# Definition of the type used for specifying the output port(s) of a processing
# unit:
#
class OutputPortSpecification:
    """
    User-defined specifications describing an output port (iterated or not).
    """

    # Predefined sequence of attributes:
    __slots__ = ('name', 'comment', 'is_iteration', 'produces_result',
                 'value_semantics', 'value_unit', 'value_type_description',
                 'value_constraints')

    # Initial definition of an output port specification:
    def __init__(self, name, comment, value_semantics, value_unit,
                 value_type_description, value_constraints=[],
                 is_iteration=False, produces_result=False):
        """
        Initial specification of all the characteristics of an output port.
        """

        self.name = name
        self.comment = comment
        self.is_iteration = is_iteration
        self.produces_result = produces_result
        self.value_semantics = value_semantics
        self.value_unit = value_unit
        self.value_type_description = value_type_description
        self.value_constraints = value_constraints

    # Encoding funtion:
    def encode(self) -> list:
        """Transforms an OutputPortSpecification instance into a list of pairs,
        more easily handled by ErlPort.
        """

        encoded_semantics = [semstring.encode() for semstring
                             in self.value_semantics]

        return [(b'name', self.name.encode()),
                (b'comment', self.comment.encode()),
                (b'is_iteration', self.is_iteration),
                (b'produces_result', self.produces_result),
                (b'value_semantics', encoded_semantics),
                (b'value_unit', self.value_unit.encode()),
                (b'value_type_description',
                 self.value_type_description.encode()),
                (b'value_constraints',
                 [constraint.encode() for constraint in
                  self.value_constraints])]


# Base class for input and output ports:
#
class GenericPort(TraceEmitter):
    """
    Abstract, base class describing any kind of ports (input or output) of a
    dataflow block.
    """

    # Fixed set of attributes of an input port:
    __slots__ = ('name', 'status', 'value')

    # Constructor of an input port:
    def __init__(self, name, status='unset', value=None):
        """Contructs a (generic, abstract) port.
        """

        # Stores the name of that port (also available in the keys of the input
        # or output port dictionary):
        #
        self.name = name

        # Possibly using default values then:
        self.status = status
        self.value = value

    # Updates the status of a port with a status received from Sim-Diasca:
    def update(self, encoded_status):
        """Updates the status of this port, from the counterpart information
        received from the engine.
        """

        # The status received from Erlang is either the 'unset' binary or a
        # pair in the form (in Erlang): { 'set', Value }

        if encoded_status == b'unset':
            self.status = 'unset'
            self.value = None

        elif (isinstance(encoded_status, tuple) and
              len(encoded_status) == 2 and
              encoded_status[0] == b'set'):
            self.status = 'set'
            self.value = encoded_status[1]

        else:
            self.send_error("Invalid port status received from the engine "
                            "while updating: {}".format(encoded_status))


# Definition of an input port, actually a subset of its Erlang counterpart,
# made by taking only the data relevant for the computations of an
# 'activate' method:
#
class InputPort(GenericPort):
    """
    Definition of an input port.
    """

    # This is actually a subset of its Erlang counterpart, made by taking only
    # the data relevant for the computations of the activate/1 (Python) method:

    __slots__ = ()

    # Constructor of an input port:
    def __init__(self, name, status='unset', value=None):
        """Contructs an input port.
        """
        super().__init__(name, status, value)

    def __str__(self):
        """Returns a textual representation of this input port.
        """
        return "Input port named '" + self.name + "'"


# Definition of an output port, actually a subset of its Erlang counterpart,
# made by taking only the data relevant for the computations of an 'activate'
# method:
#
class OutputPort(GenericPort):
    """
    Output port of a Python dataflow block.
    """

    __slots__ = ()

    # Constructor of an output port:
    def __init__(self, name, status='unset', value=None):
        """
        Contructor of an output port.
        """
        super().__init__(name, status, value)

    # Representation of an output port:
    def __repr__(self):
        """
        Representation of an output port.
        """
        return 'OutputPort : ' + self.name


# Base class for input and output port iterations:
#
class GenericPortIteration(TraceEmitter):
    """
    Generic port iteration of a Python dataflow block.
    """

    # Fixed set of attributes of a port iteration:
    __slots__ = ('base_name', 'port_count', 'bounds', 'indexes')

    # Constructor of a generic port iteration:
    def __init__(self, base_name, iteration_spec, indexes=None):
        """
        Contructor of a generic port iteration.
        """

        # Saves the name of the port iteration, which is the base name for all
        # its members:
        self.base_name = base_name

        # Decodes the user-defined multiplicity specification:
        if isinstance(iteration_spec, tuple):
            self.port_count = iteration_spec[0]
            bound_spec = iteration_spec[1]
            if isinstance(bound_spec, tuple):
                self.bounds = bound_spec
            else:
                self.bounds = (0, bound_spec)
        elif isinstance(iteration_spec, bool):
            self.port_count = 0
            self.bounds = (0, None)
        elif isinstance(iteration_spec, int):
            self.port_count = iteration_spec
            self.bounds = (0, None)
        else:
            self.port_count = 0
            self.bounds = (0, None)

        # List of port indexes:
        self.indexes = indexes

        if indexes is not None and self.port_count != len(indexes):
            self.send_error("Initially, the list of indexes (length {}) "
                            "does not match the declared multiplicity ({}) "
                            "for the input port iteration {}.".format(
                                len(indexes), self.port_count, base_name))

    # Updates the port iteration according to the data received from Erlang:
    def update(self, erlang_multiplicity, erlang_indexes: List[int]):
        """
        Updates the port iteration according to the data received from Erlang.
        """

        # In the Erlang part of Sim-Diasca, multiplicity is defined as a tuple
        # of length 2 of the form { port_count, { min_bound, max_bound } }:
        if (isinstance(erlang_multiplicity, tuple) and
                len(erlang_multiplicity) == 2):

            # The count of member ports is just copied:
            self.port_count = erlang_multiplicity[0]

            # The bounds are stored as tuple of length 2 which is also copied:
            bounds_tuple = erlang_multiplicity[1]
            if (isinstance(bounds_tuple, tuple) and
                    len(bounds_tuple) == 2):
                self.bounds = bounds_tuple

            else:
                self.send_error("Ill-formed bounds tuple received from "
                                "Erlang for the port iteration '{}': {}"
                                .format(self.base_name, bounds_tuple))

        else:
            self.send_error("Ill-formed multiplicity received from Erlang "
                            "while updating the port iteration '{}': {}".format(
                                self.base_name, erlang_multiplicity))

        # The port indexes are a simple list being copied:
        self.indexes = erlang_indexes


# Definition of an input port, actually a subset of its Erlang counterpart,
# made by taking only the data relevant for the computations of an
# 'activate' method:
#
class InputPortIteration(GenericPortIteration):
    """
    Input port iteration of a Python dataflow block.
    """

    __slots__ = ()

    # Constructor of an input port iteration:
    def __init__(self, base_name, iteration_spec, indexes=None):
        """
        Contructor of an input port iteration.
        """
        super().__init__(base_name, iteration_spec, indexes)

    # Representation of an input port iteration:
    def __repr__(self):
        """
        Representation of an input port iteration.
        """
        return 'InputPortIteration : ' + self.base_name


# Definition of an input port, actually a subset of its Erlang counterpart,
# made by taking only the data relevant for the computations of an
# 'activate' method:
#
class OutputPortIteration(GenericPortIteration):
    """
    Output port iteration of a Python dataflow block.
    """

    __slots__ = ()

    # Constructor of an output port iteration:
    def __init__(self, base_name, iteration_spec, indexes=None):
        """
        Contructor of an output port iteration.
        """
        super().__init__(base_name, iteration_spec, indexes)

    # Representation of an output port iteration:
    def __repr__(self):
        """
        Representation of an output port iteration.
        """
        return 'OutputPortIteration : ' + self.base_name


# Definition of a channel value, which is an enrichment of the simple value
# borne by a port: just aside the value, inside a tuple, we join some metadata
# used by Sim-Diasca for checking that inter-block communications make sense.
# This tuple containing a value and its metadata is called a channel value.
#
class ChannelValue:
    """
    Class representing a Python alias of the kind of values that are expected
    in all channels of a dataflow by Sim-Diasca.
    """

    __slots__ = ('value', 'value_semantics', 'value_unit', 'value_type')

    def __init__(self, value, value_semantics: List[str], value_unit: str,
                 value_type: str):
        """
        Creates a new channel value from a basic value and some metadata: the
        semantics revealing what the value represents to humans, the physical
        unit with which it should be accompanied and its computational type.
        """

        self.value = value
        self.value_semantics = value_semantics
        self.value_unit = value_unit
        self.value_type = value_type

    def encode(self):
        """
        Encodes the metadata and zips the basic value with them in a 4-tuple.
        """

        encoded_semantics = [semstring.encode() for semstring
                             in self.value_semantics]

        return (self.value, encoded_semantics, self.value_unit.encode(),
                self.value_type.encode())

    def __repr__(self):
        """
        Textual representation of this channel value
        """
        return 'ChannelValue {} {} typed {} with semantics {}'.format(
            self.value, self.value_unit, self.value_type,
            self.value_semantics)

##
##
# --- DERIVED TYPES ---
##
##

PortName = str

BlockName = str

File: sim-diasca/src/core/services/dataflow/bindings/java/api/sim_diasca/DataflowTypes.java
/**
 * 
 */
package sim_diasca;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.ericsson.otp.erlang.OtpErlangInt;
import com.ericsson.otp.erlang.OtpErlangList;
import com.ericsson.otp.erlang.OtpErlangObject;
import com.ericsson.otp.erlang.OtpErlangRangeException;
import com.ericsson.otp.erlang.OtpErlangTuple;

import common.TraceEmitter;

/**
 * @author J.Guillemineau - julien-externe.guillemineau@edf.fr
 *
 */
public class DataflowTypes {
	
	public Class[] types = new Class[]{String.class, String.class};
	
	/**
	 * Enumeration matching the activation policies available in Sim-Diasca.
	 * The activation policy governing the condition(s) to match before triggering
	 * the evaluation of the 'activate' method of a processing unit.
	 */
	public enum ActivationPolicy {
		ACTIVATE_ON_NEW_SET,
		ACTIVATE_WHEN_ALL_SET,
		CUSTOM_ACTIVATION,
	}
	
	/**
	 * Enumeration matching the constraints available in Sim-Diasca.
	 * The available constraints, against which any port value can be checked:
	 * Note: these must match the list of constraints in Sim-Diasca, which can be
	 * found in class_DataflowBlock_functions.hrl:validate_constraints/2 as Erlang atoms.
	 */
	public enum PortConstraints {
		NON_NULL,
		STRICTLY_NEGATIVE,
		NEGATIVE,
		STRICTLY_POSITIVE,
		POSITIVE,
		IN,					// to be used as ( IN, list )
		BETWEEN,			// to be used as ( BETWEEN, x, y )
		LOWER_THAN,			// to be used as ( LOWER_THAN, x )
		GREATER_THAN, 		// to be used as ( GREATER_THAN, x )
	}
	
	/**
	 * User-defined specifications describing an input port (iterated or not).
	 *
	 */
	public class InputPortsSpecification {
		
		private String name;
		private String comment;
		private Boolean is_iteration;
		private String value_semantics;
		private String value_unit;
		private String value_type_description;
		private List<PortConstraints> value_constraints;		
		
		/**
		 * Initial specification of all the characteristics of an input port.
		 * @param name
		 * @param comment
		 * @param is_iteration
		 * @param value_semantics
		 * @param value_unit
		 * @param value_type_description
		 * @param value_constraints
		 */
		public InputPortsSpecification(String name, String comment, Boolean is_iteration, String value_semantics,
				String value_unit, String value_type_description, List<PortConstraints> value_constraints) {
			super();
			this.name = name;
			this.comment = comment;
			this.is_iteration = is_iteration;
			this.value_semantics = value_semantics;
			this.value_unit = value_unit;
			this.value_type_description = value_type_description;
			this.value_constraints = value_constraints;
		}
		
		/**
		 * Initial specification of all the characteristics of an input port.
		 * @param name
		 * @param comment
		 * @param value_semantics
		 * @param value_unit
		 * @param value_type_description
		 * @param value_constraints
		 */
		public InputPortsSpecification(String name, String comment, String value_semantics, String value_unit,
				String value_type_description, List<PortConstraints> value_constraints) {
			super();
			this.name = name;
			this.comment = comment;
			this.is_iteration = false;
			this.value_semantics = value_semantics;
			this.value_unit = value_unit;
			this.value_type_description = value_type_description;
			this.value_constraints = value_constraints;
		}

		// Get/Set
		public String getName() {
			return name;
		}
		public void setName(String name) {
			this.name = name;
		}
		public Boolean getIs_iteration() {
			return is_iteration;
		}
		public void setIs_iteration(Boolean is_iteration) {
			this.is_iteration = is_iteration;
		}

		/**
		 * Transforms an InputPortSpecification object map.
		 */
		public Map<String, Object> encode() {
			Map<String, Object> map = new HashMap<>();
			map.put("name", this.name);
			map.put("comment", this.comment);
			map.put("is_iteration", this.is_iteration);
			map.put("value_semantics", this.value_semantics);
			map.put("value_unit", this.value_unit);
			map.put("value_type_description", this.value_type_description);
			map.put("value_constraints", this.value_constraints);
			return map;
		}
	}
	
	/**
	 * User-defined specifications describing an input port (iterated or not).
	 *
	 */
	public class OutputPortsSpecification {
		
		private String name;
		private String comment;
		private Boolean is_iteration;
		private Boolean produces_result;
		private String value_semantics;
		private String value_unit;
		private String value_type_description;
		private List<PortConstraints> value_constraints;		
		
		// Get/Set		
		public String getName() {
			return name;
		}
		public void setName(String name) {
			this.name = name;
		}
		public Boolean getIs_iteration() {
			return is_iteration;
		}
		public void setIs_iteration(Boolean is_iteration) {
			this.is_iteration = is_iteration;
		}

		/**
		 * Initial specification of all the characteristics of an output port.
		 * @param name
		 * @param comment
		 * @param is_iteration
		 * @param produces_result
		 * @param value_semantics
		 * @param value_unit
		 * @param value_type_description
		 * @param value_constraints
		 */
		public OutputPortsSpecification(String name, String comment, Boolean is_iteration, Boolean produces_result,
				String value_semantics, String value_unit, String value_type_description,
				List<PortConstraints> value_constraints) {
			super();
			this.name = name;
			this.comment = comment;
			this.is_iteration = is_iteration;
			this.produces_result = produces_result;
			this.value_semantics = value_semantics;
			this.value_unit = value_unit;
			this.value_type_description = value_type_description;
			this.value_constraints = value_constraints;
		}

		/**
		 * Initial specification of all the characteristics of an output port.
		 * @param name
		 * @param comment
		 * @param value_semantics
		 * @param produces_result
		 * @param value_unit
		 * @param value_type_description
		 * @param value_constraints
		 */
		public OutputPortsSpecification(String name, String comment, String value_semantics, Boolean produces_result,
				String value_unit, String value_type_description, List<PortConstraints> value_constraints) {
			super();
			this.name = name;
			this.comment = comment;
			this.produces_result = produces_result;
			this.value_semantics = value_semantics;
			this.value_unit = value_unit;
			this.value_type_description = value_type_description;
			this.value_constraints = value_constraints;
		}

		/**
		 * Initial specification of all the characteristics of an output port.
		 * @param name
		 * @param comment
		 * @param is_iteration
		 * @param value_semantics
		 * @param value_unit
		 * @param value_type_description
		 * @param value_constraints
		 */
		public OutputPortsSpecification(String name, String comment, Boolean is_iteration, String value_semantics,
				String value_unit, String value_type_description, List<PortConstraints> value_constraints) {
			super();
			this.name = name;
			this.comment = comment;
			this.is_iteration = is_iteration;
			this.value_semantics = value_semantics;
			this.value_unit = value_unit;
			this.value_type_description = value_type_description;
			this.value_constraints = value_constraints;
		}

		/**
		 * Initial specification of all the characteristics of an output port.
		 * @param name
		 * @param comment
		 * @param value_semantics
		 * @param value_unit
		 * @param value_type_description
		 * @param value_constraints
		 */
		public OutputPortsSpecification(String name, String comment, String value_semantics, String value_unit,
				String value_type_description, List<PortConstraints> value_constraints) {
			super();
			this.name = name;
			this.comment = comment;
			this.value_semantics = value_semantics;
			this.value_unit = value_unit;
			this.value_type_description = value_type_description;
			this.value_constraints = value_constraints;
		}

		/**
		 * Transforms an OutputPortSpecification object map.
		 */
		public Map<String, Object> encode() {
			Map<String, Object> map = new HashMap<>();
			map.put("name", this.name);
			map.put("comment", this.comment);
			map.put("produces_result", this.produces_result);
			map.put("is_iteration", this.is_iteration);
			map.put("value_semantics", this.value_semantics);
			map.put("value_unit", this.value_unit);
			map.put("value_type_description", this.value_type_description);
			map.put("value_constraints", this.value_constraints);
			return map;
		}
	}
	
	/**
	 * Base class for input and output ports
	 */
	public class GenericPort extends TraceEmitter {
		
		protected String name;
		protected String status;
		protected String value;		

		/**
		 * Constructor of a generic port.
		 * @param name
		 * @param status
		 * @param value
		 */
		public GenericPort(String name, String status, String value) {
			super(communicator);
			this.name = name;
			this.status = status;
			this.value = value;
		}

		/**
		 * Constructor of a generic port.
		 * @param name
		 * @param status
		 */
		public GenericPort(String name, String status) {
			super(communicator);
			this.name = name;
			this.status = status;
			this.value = null;
		}
		
		/**
		 * Constructor of a generic port.
		 * @param name
		 */
		public GenericPort(String name) {
			super(communicator);
			this.name = name;
			this.status = null;
			this.value = null;
		}
		
		// Get/set		
		public String getName() {
			return name;
		}
		public void setName(String name) {
			this.name = name;
		}

		/**
		 * Updates the status of a port with a status received from Sim-Diasca.
		 * @param encoded_status
		 */
		public void update(String[] encoded_status) {
			if(encoded_status[0].equals("unset")) {
				this.status = "unset";
				this.value = null;
			} else if (encoded_status.length == 2 && encoded_status[0].equals("set")) {
				this.status = "set";
				this.value = encoded_status[1];
			} else {
				this.sendError(String.format("Bad input port status received from Erlang while updating: %s", 
						encoded_status.toString()));
			}
		}
	}
	
	/**
	 * Definition of an input port, actually a subset of its Erlang counterpart, made by taking only 
	 * the data relevant for the computations of an 'activate' method
	 */
	public class InputPort extends GenericPort {

		public InputPort(String name, String status, String value) {
			super(name, status, value);
		}
		public InputPort(String name, String status) {
			super(name, status, null);
		}
		public InputPort(String name) {
			super(name);
		}
		@Override
		public String toString() {
			return String.format("InputPort : %", this.name);
		}		
	}
	
	/**
	 * Definition of an output port, actually a subset of its Erlang counterpart, made by taking only 
	 * the data relevant for the computations of an 'activate' method
	 */
	public class OutputPort extends GenericPort {

		public OutputPort(String name, String status, String value) {
			super(name, status, value);
		}
		public OutputPort(String name, String status) {
			super(name, status, null);
		}
		public OutputPort(String name) {
			super(name);
		}
		@Override
		public String toString() {
			return String.format("OutputPort : %", this.name);
		}		
	}
	
	/**
	 * Base class for input and output port iterations
	 */
	public class GenericPortIteration extends TraceEmitter {

		protected String base_name;
		private int port_count;
		private OtpErlangTuple bounds;
		private OtpErlangList indexes = null;
		private OtpErlangObject iteration_spec;
		
		/**
		 * Constructor of a generic port iteration.
		 * @param communicator
		 * @param base_name
		 * @param iteration_spec
		 * @param indexes
		 * @throws OtpErlangRangeException 
		 */
		public GenericPortIteration(String base_name, OtpErlangObject iteration_spec, OtpErlangList indexes) 
				throws OtpErlangRangeException {
			super(communicator);
			this.base_name = base_name;
			this.indexes = indexes;
			this.iteration_spec = iteration_spec;
			
			if(iteration_spec instanceof OtpErlangTuple) {
				OtpErlangTuple tuple_iteration_spec = (OtpErlangTuple) iteration_spec;
				this.port_count = ((OtpErlangInt)tuple_iteration_spec.elementAt(0)).intValue();
				if(tuple_iteration_spec.elementAt(1) instanceof OtpErlangTuple) {
					this.bounds = (OtpErlangTuple)tuple_iteration_spec.elementAt(1);
				} else {
					this.bounds = new OtpErlangTuple(new OtpErlangObject[] {new OtpErlangInt(0), 
							new OtpErlangTuple(tuple_iteration_spec).elementAt(1)});
				}
			} else if (iteration_spec instanceof OtpErlangInt) {
				this.port_count = ((OtpErlangInt)iteration_spec).intValue();
				this.bounds = new OtpErlangTuple(new OtpErlangObject[] {new OtpErlangInt(0), null});
			} else {
				this.port_count = 0;
				this.bounds = new OtpErlangTuple(new OtpErlangObject[] {new OtpErlangInt(0), null});
			}		

			this.indexes = indexes;
				
			if(this.indexes != null && this.port_count != this.indexes.arity()) {
				this.sendError(String.format("Initially, the list of indexes (length %s) does not match "
						+ "the declared multiplicity (%s) for the input port iteration %s.", 
						this.indexes.arity(), this.port_count, this.base_name));
			}			
		}
		
		// Get/Set
		public OtpErlangList getIndexes() {
			return indexes;
		}
		public void setIndexes(OtpErlangList indexes) {
			this.indexes = indexes;
		}				
		public int getPort_count() {
			return port_count;
		}
		public void setPort_count(int port_count) {
			this.port_count = port_count;
		}

		/**
		 * Updates the port iteration according to the data received from Erlang
		 * @param erlang_multiplicity
		 * @param erlang_indexes
		 * @throws OtpErlangRangeException 
		 */
		public void update(OtpErlangObject multiplicity, OtpErlangList list_indexes) throws OtpErlangRangeException {
			
			if(multiplicity instanceof OtpErlangTuple && list_indexes.arity() == 2) {
				OtpErlangTuple tuple_multiplicity = (OtpErlangTuple) multiplicity;
				
				// The count of member ports is just copied
				this.port_count = ((OtpErlangInt)tuple_multiplicity.elementAt(0)).intValue();
				
				// The bounds are stored as tuple of length 2 which is also copied
				if(tuple_multiplicity.elementAt(1) instanceof OtpErlangTuple && 
						((OtpErlangTuple)tuple_multiplicity.elementAt(1)).arity() == 2) {
					OtpErlangTuple bounds_tuple = (OtpErlangTuple)tuple_multiplicity.elementAt(1);
					this.bounds = bounds_tuple;
				} else {
					this.sendError(String.format("Ill-formed bounds tuple received from Erlang for the port "
							+ "iteration %s: %s ",
							this.base_name, bounds.toString()));
				}
			} else {
				this.sendError(String.format("Ill-formed multiplicity received from Erlang while updating the "
						+ "port iteration %s: %s",
						this.base_name, multiplicity.toString()));
			}
			
			this.indexes = list_indexes;
		}
	}
	
	/**
	 * Definition of an input port, actually a subset of its Erlang counterpart, made by taking only the data relevant 
	 * for the computations of an 'activate' method
	 */
	public class InputPortIteration extends GenericPortIteration{

		public InputPortIteration(String base_name, OtpErlangObject iteration_spec, OtpErlangList indexes) 
				throws OtpErlangRangeException {
			super(base_name, iteration_spec, indexes);
		}

		@Override
		public String toString() {
			return "InputPortIteration " + this.base_name;
		}		
	}
	
	public class OutputPortIteration extends GenericPortIteration{

		public OutputPortIteration(String base_name, OtpErlangObject iteration_spec, OtpErlangList indexes) 
				throws OtpErlangRangeException {
			super(base_name, iteration_spec, indexes);
		}
		
		@Override
		public String toString() {
			return "OutputPortIteration " + this.base_name;
		}
	}
	
	/**
	 * Definition of a channel value, which is an enrichment of the simple value borne by a port: just aside the value, 
	 * inside a tuple, we join some meta data used by Sim-Diasca for checking that inter-block communications make sense.
	 * This tuple containing a value and its meta data is called a channel value.
	 */
	public class ChannelValue{
		private double value;
		private String value_semantics;
		private String value_unit;
		private String value_type;
		
		/**
		 * Creates a new channel value from a basic value and some metadata: the semantics revealing what the value
		 * represents to humans, the physical unit with which it should be accompanied and its computational type.
		 * @param actual_impact
		 * @param value_semantics
		 * @param value_unit
		 * @param value_type
		 */
		public ChannelValue(double actual_impact, String value_semantics, String value_unit, String value_type) {
			super();
			this.value = actual_impact;
			this.value_semantics = value_semantics;
			this.value_unit = value_unit;
			this.value_type = value_type;
		}				
	}
		
	public class ErlangMultiplicity {
		
		private int port_count = -1;
		private int min_bound = -1;
		private int max_bound = -1;
		
		/**
		 * @param port_count
		 * @param min_bound
		 * @param max_bound
		 */
		public ErlangMultiplicity(int port_count, int min_bound, int max_bound) {
			super();
			this.port_count = port_count;
			this.min_bound = min_bound;
			this.max_bound = max_bound;
		}

		// Get/Set
		public int getPort_count() {
			return port_count;
		}
		public void setPort_count(int port_count) {
			this.port_count = port_count;
		}
		public int getMin_bound() {
			return min_bound;
		}
		public void setMin_bound(int min_bound) {
			this.min_bound = min_bound;
		}
		public int getMax_bound() {
			return max_bound;
		}
		public void setMax_bound(int max_bound) {
			this.max_bound = max_bound;
		}	
		
		@Override
		public String toString() {
			return "ErlangMultiplicity [port_count=" + port_count + ", min_bound=" + min_bound + ", max_bound="
					+ max_bound + "]";
		}
	}
}

